### ISSUE TITLE
[Model Bringup] Gemma4-12b model

### ISSUE THREAD

**Issue by saiarthiraguram** [tenstorrent/tt-xla#5255]
# google/gemma-4-12B on Tenstorrent — Bringup Documentation

**Status:** text-only single-forward path **verified on-device** (TP=8, n300-llmbox, bf16, real weights).
Autoregressive generation, multimodal (image/video/audio), and performance benchmarking are **scoped but not yet brought up** 

---

## 1. Model facts (from `config.json`, no work needed — cite)

| Field | Value |
|-------|-------|
| HF model ID | `google/gemma-4-12B` (released 2026-06-04, base / **non-instruct**) |
| Architecture | `Gemma4UnifiedForConditionalGeneration`, `model_type=gemma4_unified` |
| Modality | any-to-any (text + vision + audio); **only the text path is brought up** |
| Parameters | 11,959,730,176 (≈11.96B, full unified incl. vision+audio towers) |
| Dtype | bfloat16 (safetensors = 23 GB on disk) |
| Forward signature | `(input_ids, pixel_values, pixel_values_videos, input_features, attention_mask, …)` |
| transformers required | `5.10.0.dev0` (git main — unreleased; provides `gemma4_unified`) |

### Text decoder (`text_config`)
| Field | Value |
|-------|-------|
| **Max context (`max_position_embeddings`)** | **262,144 tokens (256K)** |
| `vocab_size` | **262,144** — this is the `262144` in the golden output `(1, 7, 262144)`, **NOT** seq len |
| Layers | 48 total = **40 `sliding_attention` + 8 `full_attention`** |
| `sliding_window` | 1024 (interleaved sliding-window attention) |
| hidden_size | 3840 · intermediate_size 15360 |
| heads | 16 attention heads, 8 KV heads, head_dim 256 |
| `attention_k_eq_v` | True (full-attention layers reuse key as value → `v_proj is None`) |
| `num_global_key_value_heads` | 1 (single unshardable global KV head on the 8 full layers) |

### Generation defaults (`generation_config.json`)
`do_sample=True`, `temperature=1.0`, `top_k=64`, `top_p=0.95`, `bos=2`, `eos=1`, `pad=0`,
`suppress_tokens=[258883, 258882]` (image-boundary tokens).

### Multimodal towers (present in config, not brought up)
`vision_config` (`gemma4_unified_vision`, patch_size 16), `audio_config` (`gemma4_unified_audio`, hidden 640).
Token IDs: `image_token_id=258880`, `video_token_id=258884`, `audio_token_id=258881`,
`boi=255999`, `eoi=258882`, `boa=256000`.

---

### Reproduce the verified on-device run
```bash
cd /proj_sw/user_dev/ctr-sraguram/tt-xla
source venv/activate
export HF_HOME=/proj_sw/user_dev/ctr-sraguram/hf_cache   # offline weights + tokenizer
export TT_METAL_CACHE=/proj_sw/user_dev/ctr-sraguram/tt_metal_cache  # JIT kernel cache OFF tiny $HOME
export TT_XLA_ARCH=n300-llmbox
export TT_VISIBLE_DEVICES=0,1,2,3      # 4 resettable boards (tt-smi); 8 runtime chips drive the mesh
export TT_XLA_SPMD=1
export CONVERT_SHLO_TO_SHARDY=1
python -m pytest -svv \
  "tests/runner/test_models.py::test_all_models_torch[gemma4/pytorch-12B-tensor_parallel-inference]"
```

---

## 3. Verified path (text-only forward, TP=8)

- **Test:** `test_all_models_torch[gemma4/pytorch-12B-tensor_parallel-inference]`
- **Mesh:** `(1, 8)` — Megatron-style 1D TP across 8 chips. `Created device mesh: (1, 8) with 8 devices.`
- **Sharding** (`loader.load_shard_spec`): column-parallel `q_proj`, MLP `gate/up`; row-parallel
  `o_proj`, MLP `down`. **KV projections replicated** (the 8 full-attention layers have
  `v_proj=None` and a single global KV head that cannot be split — standard GQA-TP fallback).
  Embeddings / lm_head / norms / vision+audio towers replicated.
- **Input:** `{input_ids, attention_mask}`, `(1, 7)` — `"What is your favorite city?"`, text path only
  (`pixel_values` forced `None` by passing inputs by keyword, not position).
- **Output:** `Gemma4UnifiedCausalLMOutputWithPast` → `.logits` `(1, 7, 262144)`.
- **History:** iter3 (1 passed, 358.65s), iter4 (1 passed, 227.56s) on a **prior reservation**;
  **iter8 (1 passed, 402.64s) on the current reservation (2026-06-17)** — the 8-chip fabric
  **recovered** after a `tt-smi -r` reset (the iter5–7 faults cleared; all 8 chips now resolve under
  SPMD and the full text path re-verified PCC-clean). Log: `logs/iter8_verify_recovery.log`.
- **`use_cache=False`** is forced on the config — the on-device path is a single full-sequence
  forward, **not** an autoregressive decode (see §6.2).

---

## 4. Sequence length - https://github.com/tenstorrent/tt-xla/issues/5255#issuecomment-4729507258

**Documented facts (above):** max context **262,144 (256K)**; sliding window 1024 over 40 of 48 layers;
the `262144` in the output is vocab, not seq len.

---

## 5. Modalities support table 

| Modality | Config support | Brought up on-device? | Notes |
|----------|----------------|------------------------|-------|
| **Text** | ✅ | ✅ **TP=8, PCC-verified** | logs documented here |
| Image + text | ✅ (`vision_config`, `image_token_id`) | ❌ | loader forces `pixel_values=None`; vision tower **replicated**, not sharded; needs new shard spec + real `pixel_values` capture |
| Video | ✅ (`pixel_values_videos`, `video_token_id`) | ❌ | frame format TBD; same tower work as image |
| Audio | ✅ (`audio_config`, `input_features`, `audio_token_id`) | ❌ | feature/mel format TBD; audio tower replicated |

---

## 6. On-device text generation  — **COMPLETED 2026-06-17, real weights, TP=8**

A standalone autoregressive harness (`gen_perf_harness.py`, this dir) was built and **run on-device**
on the n300-llmbox at TP=8 (mesh (1,8), bf16), reusing the **exact verified shard spec**
(`loader.load_shard_spec`). Results: `gen_perf_results_greedy.{json,txt}` (greedy) and
`gen_perf_results_sample.{json,txt}` (sampling, `temperature=1.0, top_k=64, top_p=0.95, seed=0` —
the model's `generation_config` defaults). Logs: `logs/gen_perf_greedy.log`, `logs/gen_perf_sample.log`.

**Decode design (important):** the loader forces `use_cache=False`, so there is **no KV cache**.
The harness keeps the input shape **constant** at `(1, buffer_len=64)` — tokens are written into a
fixed buffer and `attention_mask` gates the live prefix — so the on-device graph is identical every
step and compiles **once**. Consequence: every step (prefill *and* each decode token) is a full
`buffer_len` forward, so prefill latency == decode-step latency. This is the honest no-KV-cache
number for this bringup.

**Greedy outputs (16 new tokens each, base/non-instruct → plain text, no chat template):**

| Prompt type | Prompt | Generated (greedy) | Verdict |
|-------------|--------|--------------------|---------|
| factual | "The three primary colors are" | " red, blue, and yellow. These colors cannot be created by mixing other colors" | ✅ correct |
| few-shot | "…English: house => French:" | " maison" (then drifts) | ✅ correct token |
| code | "def fib(n):" | "    if n == 0:\n        return 0\n    elif" | ✅ plausible |
| reasoning | "3 apples, take away 2…" | " 1. If there are 3 apples and you take away 2," | ⚠️ base-model drift |
| summarization | "Summarize…Eiffel Tower…" | " The Eiffel Tower is a…" (repeats) | ⚠️ degenerate (greedy) |
| non-English | "Übersetze ins Englische: 'Guten Morgen…'" | " 'Good morning, how are you?'" | ✅ correct |
| short Q&A | "What is the capital of France?" | (empty / newlines) | ⚠️ base model, no instruct |

**Sampling** (same prompts) gave more varied, often cleaner output — e.g. code →
`if n < 2:\n    return n\nelse:`; summarization echoed the source fact; non-English correct.
Quality is exactly what a **base (non-instruct)** checkpoint produces; an instruct variant would
sharpen Q&A/summarization. ✅ greedy **and** sampling paths both work on-device.

## 7. Performance — **COMPLETED 2026-06-17, on-device TP=8 bf16**

All numbers: **arch n300-llmbox, TP=8 mesh (1,8), dtype bf16, batch 1, buffer_len 64**.

| Metric | Value | Notes |
|--------|-------|-------|
| One-time compile (first (1,64) forward) | **~319 s cold** / **~133 s warm** | cold = empty `TT_METAL_CACHE`; warm = kernel cache populated (dynamo+stablehlo still re-lower per process) |
| Steady-state forward latency | **~0.48–0.49 s** | one full (1,64) forward, compile amortized |
| **TTFT / prefill** (compile amortized) | **~0.46–0.50 s** | == one forward (no KV cache) |
| **Decode throughput** | **~2.0–2.2 tokens/s** | == 1 / forward latency (every step is a full forward, no KV cache) |
| End-to-end, 16 new tokens (warm) | **~7.8 s** | 16 × ~0.49 s |
| Model load (HF → torch, bf16) | ~2.7–19.7 s | varies with FS cache warmth |
| Shard + place on mesh | ~4.4–13.8 s | `to(device)` + 240 `mark_sharding` calls |
| Peak host RSS | **~26 GiB** | weights lazily materialized on device; host stays modest |

**Caveats / not captured:** throughput-vs-batch (batch=1 only), peak *device* memory (host RSS only;
device-mem needs runtime instrumentation), and latency-vs-seq-len (ties to the §1 256K sweep, which the
user de-prioritized). The headline `~2 tok/s` is **KV-cache-disabled** — a static-cache decode path
(mistral-style `StaticCache` + `cache_position`, sharded across the mesh) would raise this
substantially but was out of scope for the text bringup.

### 8.1 Multimodal 
Each non-text tower needs: shard spec (towers currently replicated), real input capture
(`pixel_values` / `pixel_values_videos` / `input_features`), and a forward that stays on the
multimodal path. **Effort: large**, one tower at a time.

---

## Logs

[gen_perf_greedy.log](https://github.com/user-attachments/files/29042193/gen_perf_greedy.log)
[gen_perf_sample.log](https://github.com/user-attachments/files/29042192/gen_perf_sample.log)
[iter8_verify_recovery.log](https://github.com/user-attachments/files/29042191/iter8_verify_recovery.log)
[bringup_steps.txt](https://github.com/user-attachments/files/29042199/bringup_steps.txt)

**Comment by saiarthiraguram** [comment]
Seq-len details:
**Arch:** n300-llmbox (8× Wormhole_b0) · **Mesh:** (1, 8) Megatron 1D TP ·
**Dtype:** bf16 · **Batch:** 1 · **Real weights** (677 shards, 23 GB).
Harness: `gen_perf_harness.py` · raw data: `seqlen_sweep/summary.csv` + `seqlen_sweep/n*.log`.

---

## Findings

- **Max usable sequence length on the current path = 2048 tokens. 4096 OOMs.**
- The OOM ceiling is set by the **dense `lm_head` logits over the full sequence**
  (`N × 262144 × 2 B`), **not** the transformer decoder and **not** the model's 256K context window.
- It is therefore a **harness/setup artifact**: a generation-shaped path that runs `lm_head` on only
  the last token removes the dominant buffer and should lift the usable context far beyond 2K.

---

## Config facts (from `config.json`, not measured)

| Field | Value |
|-------|-------|
| `max_position_embeddings` (context window) | **262,144 (256K)** |
| `vocab_size` | **262,144** — the `262144` in the output `(1, N, 262144)` is **vocab, not seq len** |
| Attention | 48 layers = **40 `sliding_attention` (window 1024) + 8 `full_attention`** |
| `sliding_window` | 1024 |
| heads | 16 attention / 8 KV / head_dim 256 · hidden 3840 |

---

## What was measured

In `gen_perf_harness.py` the on-device input is a **fixed `(1, buffer_len)` buffer** and the forward
runs over the **whole** buffer every step (the attention mask only gates the live prefix). So
**`buffer_len` is the exercised sequence length**: RoPE positions span `arange(buffer_len)` and the
8 full-attention layers compute the full `N×N` attention. Each `N` is a distinct compiled shape
(~170–190 s cold compile); per `N` we ran one prefill forward + one decode-step forward (the decode
step, on the cached shape, gives the pure forward latency at length `N`).

### Sweep results

| seq len `N` | result | forward latency | logits buffer `N × 262144 × 2 B` |
|-------------|--------|-----------------|----------------------------------|
| 128  | ✅ fit | 0.448 s | 64 MiB |
| 512  | ✅ fit | 0.686 s | 256 MiB |
| 1024 | ✅ fit | 1.286 s | 512 MiB |
| **2048** | ✅ **fit (max)** | 2.532 s | **1 GiB** |
| 4096 | ❌ **OOM** | — | **2 GiB** |

Forward latency is smooth and roughly linear in `N` (≈ doubles as `N` doubles). Compile time is
~constant per shape (~170–190 s).

---

## Root cause of the OOM

The 4096 failure is a single **2 GiB DRAM buffer** allocation, which is **exactly**
`4096 × 262144 × 2 B = 2,147,483,648 B` — the logits tensor `(1, 4096, 262144)` in bf16, i.e. the
`lm_head` projection computed for **every position** over the 262,144-token vocabulary:

```
TT_FATAL: Out of Memory: Not enough space to allocate 2147483648 B DRAM buffer across 12 banks,
where each bank needs to store 179306496 B, but bank size is 1071821792 B
(allocated: 797204608 B, free: 274617184 B, largest free block: 178958336 B)
```

- At **N=2048** this buffer is **1 GiB** → fits.
- At **N=4096** it is **2 GiB** → exceeds free DRAM (only ~274 MB free, largest block ~178 MB).

The decoder stack (attention + MLP) clearly has headroom at these lengths — the latency curve is
smooth with no memory cliff until the logits buffer crosses the limit. **The ceiling is the dense
vocab projection over the full sequence, nothing else.**

---
